Serialize mapped tasks and task groups

ashb opened the pull request:
This is needed for the scheduler to be able to expand the task instances at runtime

ashb committed:
Basic serialization support for MappedOperators

ashb committed:
Serialization support for TaskGroups

uranusjr commented on airflow/models/baseoperator.py:
```diff
@@ -1634,7 +1635,7 @@ def map(self, **kwargs) -> "MappedOperator":
 
 def _validate_kwarg_names_for_mapping(cls: Type[BaseOperator], func_name: str, value: Dict[str, Any]):
-    if isinstance(str, cls):
+    if isinstance(cls, str):
```
Neither the old and new line seem to make sense? Should `cls` be annotated as `Union[Type[BaseOperator], str]` instead?

ashb replied:
Yes, it should be -- I corrected that type annotation on the MappedOperator but not here.

uranusjr replied:
I pushed a commit to fix this (and a few other annotations)

uranusjr commented on airflow/models/baseoperator.py:
```diff
@@ -1663,24 +1664,64 @@ def _validate_kwarg_names_for_mapping(cls: Type[BaseOperator], func_name: str, v
+
+    for field in fields:
+        results.append(field.evolve(repr=False))
+    return results
```
This feels much too magical to me 🙁 Why not add `repr=False` explicitly instead, or implement a custom `__repr__`?

ashb replied:
Fair comment.

uranusjr commented on airflow/models/baseoperator.py:
```diff
@@ -207,6 +207,7 @@ def apply_defaults(self: "BaseOperator", *args: Any, **kwargs: Any) -> Any:
 
             result = func(self, **kwargs, default_args=default_args)
             # Store the args passed to init -- we need them to support task.map serialzation!
+            kwargs.pop('task_id', None)
```
I don’t get what this implies.

ashb replied:
Delete the task_id from the kwargs (so it isn't stored in the mapped data structure) -- so `del kwargs['task_id']` but without the try/except for when it's not there? I'll double check if I can just do this with a `del`.

uranusjr replied:
I guess my question is more like why do we need to treat `task_id` specially (and why we didn’t need to previously, but only when we add operator mapping)

ashb replied:
Task ID was getting stored in partial_kwargs which I hadn't noticed until I looked at the serialized representation of a MappedOperator.

uranusjr replied:
How about storing `task_id` separately instead and not in `partial_kwargs`?

ashb replied:
Oh right I remember the problem now. Because we end up calling apply_defaults multiple times (once for each class in the MRO) task_id gets added in to kwargs for the `super().__init__(**kwargs)` call, so this del here is to make sure it doesn't get stored in partial kwargs.

uranusjr replied:
Hmm shouldn’t this be fixed on `partial_kwargs` instead then? I’m surprised this doesn’t cause problems elsewhere. Or is `task_id` being present here wrong in the first place and should be generally fixed? (In that case we should add a comment.)

ashb replied:
This is dealing with it in partial_kwargs -- 

```
            # Store the args passed to init -- we need them to support task.map serialzation!
            kwargs.pop('task_id', None)
            self._BaseOperator__init_kwargs.update(kwargs)  # type: ignore
```

(The only use of `__init_kwargs` is for building partial_kwargs when mapping a task)

ashb replied:
This is the other option:

```diff
diff --git a/airflow/models/baseoperator.py b/airflow/models/baseoperator.py
index 752b8a6cf..71170320d 100644
--- a/airflow/models/baseoperator.py
+++ b/airflow/models/baseoperator.py
@@ -136,6 +136,7 @@ class BaseOperatorMeta(abc.ABCMeta):
         non_optional_args = {
             name for (name, param) in non_variadic_params.items() if param.default == param.empty
         }
+        non_optional_args -= {'task_id'}
 
         class autostacklevel_warn:
             def __init__(self):
@@ -158,7 +159,7 @@ class BaseOperatorMeta(abc.ABCMeta):
             func.__globals__['warnings'] = autostacklevel_warn()
 
         @functools.wraps(func)
-        def apply_defaults(self: "BaseOperator", *args: Any, **kwargs: Any) -> Any:
+        def apply_defaults(self: "BaseOperator", *args: Any, task_id: str, **kwargs: Any) -> Any:
             from airflow.models.dag import DagContext
             from airflow.utils.task_group import TaskGroupContext
 
@@ -201,15 +202,15 @@ class BaseOperatorMeta(abc.ABCMeta):
 
             hook = getattr(self, '_hook_apply_defaults', None)
             if hook:
-                args, kwargs = hook(**kwargs, default_args=default_args)
+                args, kwargs = hook(task_id=task_id, **kwargs, default_args=default_args)
+                task_id = kwargs.pop('task_id')
                 default_args = kwargs.pop('default_args', {})
 
             if not hasattr(self, '_BaseOperator__init_kwargs'):
                 self._BaseOperator__init_kwargs = {}
 
-            result = func(self, **kwargs, default_args=default_args)
+            result = func(self, **kwargs, task_id=task_id, default_args=default_args)
             # Store the args passed to init -- we need them to support task.map serialzation!
-            kwargs.pop('task_id', None)
             self._BaseOperator__init_kwargs.update(kwargs)  # type: ignore
 
             # Here we set upstream task defined by XComArgs passed to template fields of the operator

```

Which do you think is better @uranusjr?

uranusjr replied:
The “other” one looks better to me, it feels like we are intentionally treating `task_id` different from all other arguments.

ashb replied:
Done in https://github.com/apache/airflow/pull/20743/commits/9439378183cb4da70e33e9c6f04d2d52924709a5 -- and a bit of a justification/reason in the commit message

uranusjr replied:
Turns out this has an unfortunate side effect:

```pycon
>>> FooOperator()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: apply_defaults() missing 1 required keyword-only argument: 'task_id'
```

and error message seems to be hard-coded in the interpreter and cannot be patched. I’m trying yet another approach (see new commits)

uranusjr commented on airflow/models/baseoperator.py:
```diff
@@ -1741,6 +1788,28 @@ def leaves(self) -> List["MappedOperator"]:
+        # This looks like `isinstance(self, DummyOperator) would work, but this also
+        # needs to cope when `self` is a Serialized instance of a DummyOperator or one
+        # of its sub-classes (which don't inherit from anything but BaseOperator).
+        return getattr(self, '_is_dummy', False)
```
`_is_dummy` has a default, so when can it be unset?

ashb replied:
Possibly can't be, I'll check.

kaxil replied:
`SerializedBaseOperator` inherits from BaseOperator too so `_is_dummy` should be already set

uranusjr commented on airflow/serialization/serialized_objects.py:
```diff
@@ -623,7 +647,7 @@ def deserialize_operator(cls, encoded_op: Dict[str, Any]) -> BaseOperator:
         for k, v in encoded_op.items():
 
-            if k == "_downstream_task_ids":
+            if k in {"_downstream_task_ids", "downstream_task_ids"}:
```
Maybe we should rename `downstream_task_ids` to `_downstream_task_ids` instead? This feels a bit awkward.

uranusjr commented on airflow/serialization/serialized_objects.py:
```diff
@@ -653,10 +677,14 @@ def deserialize_operator(cls, encoded_op: Dict[str, Any]) -> BaseOperator:
 
-            setattr(op, k, v)
+            if hasattr(op, k) and isinstance(v, set):
+                getattr(op, k).update(v)
```
What does this cover? Too hacky IMO.

ashb replied:
```python
    @property
    def upstream_task_ids(self) -> Set[str]:
        """@property: set of ids of tasks directly upstream"""
        return self._upstream_task_ids

    @property
    def downstream_task_ids(self) -> Set[str]:
        """@property: set of ids of tasks directly downstream"""
        return self._downstream_task_ids
```

Those can't be written but to the value of the set is mutable. So it's related to the `downstream_task_ids` vs `_downstream_task_ids` -- I think the change might be to make `downstream_task_id` a normal set attribute on BaseOperator (rather than have `_downstream_task_ids` on MappedOperator) -- I don't like storing/accessing the "private" fields here.

I'll see if that can be done without breaking the old serialization format.

ashb replied:
I'f changed this slightly to:

```
try:
  setattr(op, k, k)
except AttributeError:
   # Read only value, update in place
  getattr(op, k).update(v)
```

still a bit hacky, but marginally less so. I'll revisit this in the future.

uranusjr replied:
Does this affect anything other than down/upstream_task_ids? If not, maybe we should just rename the attribute on BaseOperator to remove the `_` (removing the public property, of course), and add a special case on deserialisation to migrate those attributes instead.

ashb replied:
Changed this (at tested that we still correctly inflate the old name correctly)

uranusjr replied:
Does the change in `deserialize_operator` also need to be updated?

ashb replied:
Which change? It's now this

```
            if k == "_downstream_task_ids":
                # Upgrade from old format/name
                k = "downstream_task_ids"
```

And then the set-except-get-update has done.

uranusjr replied:
Around here: https://github.com/apache/airflow/pull/20743/files#diff-807ca0a4fd53aeb41166621c9842b0f89b7931fc64e9a60befa36c776db45efaL661-L662

ashb replied:
Ah no, can't remove that one yet. I've added a TODO comment to refactor and remove it later.

uranusjr commented on airflow/serialization/serialized_objects.py:
```diff
@@ -653,10 +677,14 @@ def deserialize_operator(cls, encoded_op: Dict[str, Any]) -> BaseOperator:
         for k in op.get_serialized_fields() - encoded_op.keys() - cls._CONSTRUCTOR_PARAMS.keys():
-            setattr(op, k, None)
+            if not hasattr(op, k):
+                setattr(op, k, None)
```
Same question for this.

uranusjr committed:
Tweak type hints to reflect reality

While keeping Mypy happy.

ashb committed:
fixup! Basic serialization support for MappedOperators

kaxil commented on airflow/models/baseoperator.py:
```diff
@@ -1672,20 +1677,52 @@ def _validate_kwarg_names_for_mapping(cls: Type[BaseOperator], func_name: str, v
+        return (
+            'MappedOperator(operator_class={self._operator_class_repr(self.operator_class)}, '
+            + 'task_id={self.task_id!r}, partial_kwargs={self.partial_kwargs!r}, '
+            + 'mapped_kwargs={self.mapped_kwargs!r}, dag={self.dag})'
```
```suggestion
            f'MappedOperator(operator_class={self._operator_class_repr(self.operator_class)}, '
            + f'task_id={self.task_id!r}, partial_kwargs={self.partial_kwargs!r}, '
            + f'mapped_kwargs={self.mapped_kwargs!r}, dag={self.dag})'
```

ashb replied:
Whoops!

kaxil commented on airflow/serialization/serialized_objects.py:
```diff
@@ -586,9 +604,24 @@ def serialize_operator(cls, op: BaseOperator) -> Dict[str, Any]:
-        op = SerializedBaseOperator(task_id=encoded_op['task_id'])
+        op: Union[BaseOperator, MappedOperator]
+        # Check if it's a mapped operator
+        if "mapped_kwargs" in encoded_op:
```
Probably the following might be a better (explicit) condition ?
```suggestion
        if encoded_op.get("_is_mapped"):
```

ashb replied:
We don't store _is_mapped currently.

kaxil replied:
The following line does, no?

https://github.com/apache/airflow/blob/3ef6268e0039f40a6380394516bdf4b587bf7a4d/airflow/serialization/serialized_objects.py#L550

ashb replied:
🤦🏻

kaxil commented on airflow/serialization/serialized_objects.py:
```diff
@@ -586,9 +604,24 @@ def serialize_operator(cls, op: BaseOperator) -> Dict[str, Any]:
+            op = MappedOperator(
+                task_id=encoded_op['task_id'],
+                dag=None,
+                operator_class='.'.join(filter(None, (encoded_op['_task_module'], encoded_op['_task_type']))),
```
Why not just the following?
```suggestion
                operator_class=f'{encoded_op['_task_module']}.{encoded_op['_task_type']}',
```

ashb replied:
I honestly don't know -- Cos module and type are both required.

kaxil commented on airflow/serialization/serialized_objects.py:
```diff
@@ -586,9 +604,24 @@ def serialize_operator(cls, op: BaseOperator) -> Dict[str, Any]:
+                dag=None,
+                operator_class='.'.join(filter(None, (encoded_op['_task_module'], encoded_op['_task_type']))),
+                # These are all re-set later
+                partial_kwargs={},
```
Shouldn't `partial_kwargs={}` be a default on the `MappedOperator` itself?

ashb replied:
No, cos in all other situations where this is constructor we need to pass a value.

kaxil commented on airflow/serialization/serialized_objects.py:
```diff
@@ -586,9 +604,24 @@ def serialize_operator(cls, op: BaseOperator) -> Dict[str, Any]:
+                # These are all re-set later
+                partial_kwargs={},
+                mapped_kwargs={},
+                deps=tuple(),
+                is_dummy=False,
+                template_fields=(),
```
Same for these three, shouldn't we have a default on `MappedOperator` itself?

ashb committed:
fixup! Basic serialization support for MappedOperators

ashb committed:
Adjust to changes on main

ashb committed:
Make downstream_task_ids a normal writeable property

It simplifies a few things.

We also deal with (and test) the old name when deserializing

ashb committed:
Handle `task_id` specially in apply_defaults

Since `task_id` is handled speically in the serialization of
MappedOperators, we don't want it duplicated in to the partial_kwargs.

ashb commented on airflow/serialization/serialized_objects.py:
```diff
@@ -658,7 +694,8 @@ def deserialize_operator(cls, encoded_op: Dict[str, Any]) -> BaseOperator:
 
         for k in op.get_serialized_fields() - encoded_op.keys() - cls._CONSTRUCTOR_PARAMS.keys():
-            setattr(op, k, None)
+            if not hasattr(op, k):
```
```suggestion
            # TODO: refactor deserialization of BaseOperator and MappedOperaotr (split it out), then check could go away.
            if not hasattr(op, k):
```

ashb committed:
Update airflow/serialization/serialized_objects.py

ashb commented on airflow/serialization/serialized_objects.py:
```diff
@@ -658,7 +694,9 @@ def deserialize_operator(cls, encoded_op: Dict[str, Any]) -> BaseOperator:
 
         for k in op.get_serialized_fields() - encoded_op.keys() - cls._CONSTRUCTOR_PARAMS.keys():
-            setattr(op, k, None)
+            # TODO: refactor deserialization of BaseOperator and MappedOperaotr (split it out), then check could go away.
```
```suggestion
            # TODO: refactor deserialization of BaseOperator and MappedOperaotr (split it out), then check
            # could go away.
```

ashb committed:
Update airflow/serialization/serialized_objects.py

uranusjr commented on airflow/models/baseoperator.py:
```diff
@@ -136,6 +136,7 @@ def _apply_defaults(cls, func: T) -> T:
         non_optional_args = {
             name for (name, param) in non_variadic_params.items() if param.default == param.empty
         }
+        non_optional_args -= {'task_id'}
```
```suggestion
        non_optional_args = {
            name
            for name, param in non_variadic_params.items()
            if param.default == param.empty
            and name != "task_id"
        }
```

should be the same?

uranusjr commented on airflow/models/baseoperator.py:
```diff
@@ -1706,25 +1701,49 @@ def _validate_kwarg_names_for_mapping(cls: Type[BaseOperator], func_name: str, v
+        return (
+            f'MappedOperator(task_type={self.task_type}, '
+            + f'task_id={self.task_id!r}, partial_kwargs={self.partial_kwargs!r}, '
+            + f'mapped_kwargs={self.mapped_kwargs!r}, dag={self.dag})'
```
```suggestion
            f'MappedOperator(task_type={self.task_type}, '
            f'task_id={self.task_id!r}, partial_kwargs={self.partial_kwargs!r}, '
            f'mapped_kwargs={self.mapped_kwargs!r}, dag={self.dag})'
```

implicit string concatenation

ashb committed:
Update airflow/models/baseoperator.py

Co-authored-by: Tzu-ping Chung <uranusjr@gmail.com>

uranusjr commented on airflow/models/baseoperator.py:
```diff
@@ -1706,25 +1701,49 @@ def _validate_kwarg_names_for_mapping(cls: Type[BaseOperator], func_name: str, v
+
+        return issubclass(self.operator_class, DummyOperator)
+
+    @deps.default
+    def _deps_from_class(self):
+        return self.operator_class.deps
```
I’m assuming many of these defaults (this, `_is_dummy`, and `template_fields`, I believe) don’t need to consider when `operator_class` is a str because in that case these values would’ve been supplied explicitly instead. (It’s kind of bad it’s designed this way but I guess that can be said for many things regarding the current serialisation implementation...)

ashb replied:
Yes, exactly that. I've already started thinking in my head about how to refactor/rearchitect the serialization and deserialization.

ashb committed:
Update airflow/models/baseoperator.py

Co-authored-by: Tzu-ping Chung <uranusjr@gmail.com>

ashb committed:
fixup! Update airflow/models/baseoperator.py

uranusjr committed:
Fix usages of the old _[up|down]stream_task_ids

uranusjr committed:
Remove task_id from kwargs only in MappedOperator
